keps/sig-node/4381-dra-structured-parameters/README.md

```diff
@@ -1099,106 +960,147 @@ server-side-apply (SSA) simpler. Patching individual list elements is not
+// For resources that are not local to a node, the node name is not set. Instead,
 // the driver may use a node selector to specify where the devices are available.
+//
+// This is an alpha type and requires enabling the DynamicResourceAllocation
+// feature gate.
 type ResourceSlice struct {
     metav1.TypeMeta
     // Standard object metadata
     // +optional
     metav1.ObjectMeta
 
+    // Contains the information published by the driver.
+    //
+    // Changing the spec automatically increments the metadata.generation number.
     Spec ResourceSliceSpec
-
-    Status ResourceSliceStatus
 }
 
+const (
+    // ResourceSliceSelectorNodeName can be used in a [metav1.ListOptions]
+    // field selector to filter based on [ResourceSliceSpec.NodeName].
+    ResourceSliceSelectorNodeName = "spec.nodeName"
+    // ResourceSliceSelectorDriver can be used in a [metav1.ListOptions]
+    // field selector to filter based on [ResourceSliceSpec.Driver].
+    ResourceSliceSelectorDriver = "spec.driver"
+)
+
+// ResourceSliceSpec contains the information published by the driver in one ResourceSlice.
 type ResourceSliceSpec struct {
-    // DriverName identifies the DRA driver providing the capacity information.
+    // Driver identifies the DRA driver providing the capacity information.
     // A field selector can be used to list only ResourceSlice
     // objects with a certain driver name.
     //
     // Must be a DNS subdomain and should end with a DNS domain owned by the
-    // vendor of the driver.
-    DriverName string
+    // vendor of the driver. This field is immutable.
+    //
+    // +required
+    Driver string
 
-    // PoolName is used to identify devices. For node-local devices, this
-    // is often the node name, but this is not required.
+    // Pool describes the pool that this ResourceSlice belongs to.
     //
-    // It must not be longer than 253 and must consist of one or more DNS sub-domains
-    // separated by slashes.
-    PoolName string
+    // +required
+    Pool ResourcePool
 
-    // NodeName identifies the node which provides the devices.
+    // NodeName identifies the node which provides the resources in this pool.
     // A field selector can be used to list only ResourceSlice
     // objects belonging to a certain node.
     //
-    // This field can be used to limit access from nodes to slices with
+    // This field can be used to limit access from nodes to ResourceSlices with
     // the same node name. It also indicates to autoscalers that adding
     // new nodes of the same type as some old node might also make new
-    // devices available.
+    // resources available.
     //
-    // NodeName and NodeSelector are mutually exclusive.
-    // If both are unset, then the devices are available in the
-    // entire cluster.
+    // Exactly one of NodeName, NodeSelector and AllNodes must be set.
+    // This field is immutable.
     //
     // +optional
+    // +oneOf=NodeSelection
```

Comment:
I'm not familar with oneOf but NodeSelection doesn't match "NodeName, NodeSelector or AllNodes". Is this a typo?

Reply:
This is not a typo: "NodeName, NodeSelector or AllNodes" are part of a fictional set called "NodeSelection".

It can be called anything, as long as it's used consistently. At least that is the plan - right now, `+oneOf` doesn't do anything.

Reply:
So is this shorthand for you to add validation to only allow one of these?

Reply:
It's part of the long-term plan to replace manual validation with automatically generated code: https://github.com/alexzielenski/enhancements/blob/declarative-validation/keps/sig-api-machinery/4153-declarative-validation/README.md#alpha

At the moment, consider it part of the documentation. It's enforced through manual validation.

Reply:
We have a semi-skunkworks project to support declarative validation, such that a tag like this (may change names, but same idea) could be used to ensure exactly one field was set without hand-writing that.  I asked for these tags as documentation in the interim.